### PR TITLE
lambdas be gone

### PR DESCRIPTION
**Pull request by kbrock**
Hello Mr. Terrible,

All this augmentation comes at a cost

Want to increase the number of emitters, but simplify them.

(I'm ok with a no, or a request to prune out a few of these.)

I have another change after this, so just let me know

--k

### CODE REVIEW

File: lib/stamp/emitters/two_digits_only.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+module Stamp
+  module Emitters
+    # Emits the given field as a two-digit number with a leading
+    # zero if necessary.
+    class TwoDigitsOnly
+      attr_reader :field, :add_zero
+
+      # @param [field] the field to be formatted (e.g. +:year+)
+      def initialize(field, add_zero = false)
```
Comment (jeremyw) [inline lib/stamp/emitters/two_digits_only.rb]:
```ruby
def initialize(field, leading_zero: false)
```

I prefer a keyword argument over an unnamed boolean. I think it's time to drop support for Ruby < 2.0, so we don't need a soft options hash.

Comment (jeremyw) [inline lib/stamp/emitters/two_digits_only.rb]: And if that's the direction we go, couldn't this just be done by the TwoDigit emitter (including the `% 100`; it _is_ called "TwoDigit" after all)?

File: lib/stamp/translator.rb [hunk 2]
```diff
@@ -79,7 +78,7 @@ def time_emitter(token)
         Emitters::AmPm.new
 
       when MERIDIAN_UPPER_REGEXP
-        Emitters::AmPm.new { |v| v.upcase }
+        Emitters::AmPm.new(['AM','PM'])
```
Comment (jeremyw) [inline lib/stamp/translator.rb]: I'm not loving this.

File: lib/stamp/translator.rb [hunk 3]
```diff
@@ -89,7 +88,7 @@ def time_emitter(token)
 
       when ONE_DIGIT_REGEXP
         # 12-hour clock without leading zero
-        Emitters::Delegate.new(:hour, &HOUR_TO_12_HOUR)
+        Emitters::TwelveHour.new(:hour)
```
Comment (jeremyw) [inline lib/stamp/translator.rb]: The TwelveHour emitter is much better. :+1:

File: lib/stamp/emitters/twelve_hour.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+module Stamp
+  module Emitters
+    # Emits the given field as a two-digit number with a leading
+    # zero if necessary.
+    class TwelveHour
+      attr_reader :field, :add_zero
+
+      # @param [field] the field to be formatted (e.g. +:month+, +:year+)
+      def initialize(field, add_zero = false)
```
Comment (jeremyw) [inline lib/stamp/emitters/twelve_hour.rb]: Same comment about using a named keyword `leading_zero`.

File: lib/stamp/emitters/twelve_hour.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        @field = field
+        @add_zero = add_zero
+      end
+
+      def format(target)
+        value = target.send(field)
+        value = ((value - 1) % 12) + 1
+        add_zero && (value < 10) ? "0#{value}" : value
```
Comment (jeremyw) [inline lib/stamp/emitters/twelve_hour.rb]: Prefer if/else over ternary.

File: lib/stamp/emitters/modifiable.rb [hunk 6]
```diff
@@ -1,9 +0,0 @@
-module Modifiable
```
Comment (jeremyw) [inline lib/stamp/emitters/modifiable.rb]: I like getting rid of this. It never felt great.

File: lib/stamp/translator.rb (hunk 2 above)
Comment (kbrock) [inline lib/stamp/translator.rb]: Then you really wouldn't have liked my alternative:

```ruby
Emitters::Lookup.new(:hour, %w(am) * 12 + %w(pm) * 12)
```
But this confuses the ambigator since `field == :hour` vs `nil` that the current AmPm does

**Comment by kbrock** [comment]
I removed am/pm to another pr.
Think these changes are agreeable enough

**Comment by kbrock** [comment]
lol. I think you wrote `value < 10 ? "0#{value}" : value`. I just copied it to the other class.

File: lib/stamp/translator.rb (hunk 3 above)
Comment (jeremyw) [inline lib/stamp/translator.rb]: This could be a constant now, to avoid using more memory than necessary.

File: lib/stamp.rb [hunk 7]
```diff
@@ -11,6 +11,7 @@
 require "stamp/emitters/string"
 require "stamp/emitters/two_digit"
 require "stamp/disambiguator"
+require "stamp/emitters/twelve_hour"
```
Comment (jeremyw) [inline lib/stamp.rb]: Sort me

File: lib/stamp/emitters/twelve_hour.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+      # @param [String|Symbol] field the field to be formatted (e.g. +:month+, +:year+)
+      # @param [Hash] options the options for output
+      # @option options [String] :leading_zero (default: false) pass true add a leading 0 to output
+      def initialize(field, options = {})
+        @field = field
+        @leading_zero = options[:leading_zero]
+      end
+
+      # @param [Date|Time|DateTime] target the date to be formatted
+      def format(target)
```
Comment (jeremyw) [inline lib/stamp/emitters/twelve_hour.rb]: I'm starting to feel like the emitters should have unit tests. There's enough going on here to be tested. Maybe in another pass.

**Comment by jeremyw** [comment]
A couple small notes. Looking forward to getting this merged!

**Comment by kbrock** [comment]
I'm turning into a pumpkin
I'll rebase this tomorrow after you merge 25 (and/or 26)
If 23 is good with you, I'll probably just roll that back into here.

If you want to write a quick example for testing an emitter, I can run with it. Otherwise, I'll throw something together

**Comment by jeremyw** [comment]
Merged the travis fixes.

If you want to take a stab at an emitter unit test with minitest, go for
it. Not sure I'll have much time today.

**Comment by kbrock** [comment]
added tests for 2.  would prefer to add more tests as a separate pr.
I can pull these out of here easily enough. this pr is starting to sprawl

File: lib/stamp/translator.rb [hunk 9]
```diff
@@ -100,7 +99,7 @@ def two_digit_hour_emitter(token)
         Emitters::TwoDigit.new(:hour)
       else
         # 12-hour clock with leading zero
-        Emitters::TwoDigit.new(:hour, &HOUR_TO_12_HOUR)
+        Emitters::TwelveHour.leading_zero
```
Comment (kbrock) [inline lib/stamp/translator.rb]: this is very different from the `TWO_DIGIT_YEAR_EMITTER` approach

Thinking about moving these all up into constants at the top and then memoizing / `leading_zero` from there.

thoughts?

Comment (jeremyw) [inline lib/stamp/translator.rb]: I'd rather just remove the constants in favor of factory methods. Maybe the constants just move to the different emitter classes and the factory methods just return the constants.

Comment (kbrock) [inline lib/stamp/translator.rb]: heh. this is creeping. I remember you once telling me to keep my PRs small.

If you want me to put that into here, just holler, but I'd prefer to create another PR with unit tests and factories for the remaining emitters.

Comment (jeremyw) [inline lib/stamp/translator.rb]: That can be a separate PR

**Comment by kbrock** [comment]
anything else to push this forward?

File: lib/stamp/emitters/am_pm.rb [hunk 10]
```diff
@@ -1,22 +1,28 @@
...
         nil
       end
+
+      @@instances = {}
+      def self.lowercase
+        @@instances[:lowercase] ||= new(%w(am pm))
+      end
+
+      def self.uppercase
+        @@instances[:upercase] ||= new(%w(AM PM))
```
Comment (jeremyw) [inline lib/stamp/emitters/am_pm.rb]: ":upercase" spelling

File: lib/stamp/emitters/am_pm.rb [hunk 11]
```diff
@@ -1,22 +1,28 @@
...
       def format(target)
-        modify(target.hour < 12 ? AM : PM)
+        @values[target.hour < 12 ? 0 : 1]
       end
 
       def field
         nil
       end
+
+      @@instances = {}
```
Comment (jeremyw) [inline lib/stamp/emitters/am_pm.rb]: I'd lazy-init the hash too. Or just make these constants.

**Comment by jeremyw** [comment]
Sorry, I've been traveling. On a plane right now. A couple small comments but this is great.

**Comment by kbrock** [comment]
thanks / lol. forgot that you were probably at railsconf.

1. yea, I like the look of the hash and lowercase emitters. but the constants are better.
2. Would prefer defining the constants at the top, but they needed to be defined after runnign the initializer. so loping at the bottom seemed the best solution there.
3. I'll put in another PR that constantizes the other emitters and tests. (then I'll get to the parser which was my original intent on coming back into this code.)

File: lib/stamp/emitters/am_pm.rb [hunk 12]
```diff
@@ -1,22 +1,22 @@
...
     class AmPm
-      include Modifiable
-
-      AM = 'am'
-      PM = 'pm'
-
-      def initialize(&block)
-        @modifier = block
+      # @param [Hash] options
+      # @option options [Boolean] :upcase (default: false) uppercase meridian
```
Comment (jeremyw) [inline lib/stamp/emitters/am_pm.rb]: I have failed you. It's "meridiem", not "meridian". http://www.grammarphobia.com/blog/2014/03/ante-meridiem-vs-antemeridian.html

Maybe we should rename Emitters::AmPm emitter to Emitters::Meridiem?

Comment (kbrock) [inline lib/stamp/emitters/am_pm.rb]: I fixed the spelling. renaming am pm is all on you.

But thanks for the grammar correction. very cool

**Comment by kbrock** [comment]
Cool. Let me know if that fixes things up.
Have another branch ready to push with the rest of these changes.
Still pondering on the `AmPm`.

File: lib/stamp/emitters/am_pm.rb [hunk 13]
```diff
@@ -1,22 +1,22 @@
...
     class AmPm
-      include Modifiable
-
-      AM = 'am'
-      PM = 'pm'
-
-      def initialize(&block)
-        @modifier = block
+      # @param [Hash] options
+      # @option options [Boolean] :upcase (default: false) uppercase meridiem
```
Comment (jeremyw) [inline lib/stamp/emitters/am_pm.rb]: Is this comment accurate?

**Comment by kbrock** [comment]
thanks. updated comment for am pm constructor
